File: netlifyfunctions/functions/createEscrow.js
const jwt = require("jsonwebtoken");
const { query } = require("./db");

const JWT_SECRET = process.env.JWT_SECRET;
const HEADERS = {
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Headers': 'Content-Type, Authorization',
  'Access-Control-Allow-Methods': 'POST, OPTIONS',
  'Content-Type': 'application/json'
};

const DEFAULT_SELLER_ID = 100;

exports.handler = async (event) => {
  if (event.httpMethod === "OPTIONS") {
    return { statusCode: 200, headers: HEADERS, body: "CORS Preflight Success" };
  }

  if (event.httpMethod !== "POST") {
    return { statusCode: 405, headers: HEADERS, body: JSON.stringify({ error: "Method not allowed" }) };
  }

  let decoded;
  try {
    const authHeader = event.headers.authorization || event.headers.Authorization;
    if (!authHeader || !authHeader.startsWith("Bearer ")) {
      return { statusCode: 401, headers: HEADERS, body: JSON.stringify({ error: "Authorization token required." }) };
    }

    const token = authHeader.substring(7);
    decoded = jwt.verify(token, JWT_SECRET);

    if (!decoded || !decoded.user_id) {
      return { statusCode: 401, headers: HEADERS, body: JSON.stringify({ error: "Invalid token payload (missing user ID)." }) };
    }

    if (decoded.role !== "buyer") {
      return { statusCode: 403, headers: HEADERS, body: JSON.stringify({ error: "Forbidden: Only buyers can create escrows." }) };
    }
  } catch (err) {
    console.error("Token validation error:", err);
    return { statusCode: 401, headers: HEADERS, body: JSON.stringify({ error: "Invalid or expired token." }) };
  }

  try {
    const { amount, description, paymentMethod } = JSON.parse(event.body);
    const buyer_id = decoded.user_id;

        
        // FIX 2: The buyer ID is pulled directly from the user_id field in the JWT
        // buyer_id now comes from request body
        // const buyer_id = decoded.user_id; 
        
        if (!amount || !description || !paymentMethod) {
            return {
                statusCode: 400,
                headers: HEADERS,
                body: JSON.stringify({ error: "Missing required fields: amount, description, or paymentMethod." }),
            };
        }

        const parsedAmount = parseFloat(amount);
        if (isNaN(parsedAmount) || parsedAmount <= 0) {
            return {
                statusCode: 400,
                headers: HEADERS,
                body: JSON.stringify({ error: "Invalid amount provided." }),
            };
        }
        
        const initialStatus = 'PENDING_SELLER_ACCEPTANCE'; 

        const insertQuery = `
            INSERT INTO escrows (buyer_id, seller_id, amount, description, payment_method, status, created_at)
            VALUES ($1, $2, $3, $4, $5, $6, NOW())
            RETURNING *;
        `;

        const values = [
            parseInt(buyer_id, 10), // buyer_id is now the corrected decoded.user_id
            DEFAULT_SELLER_ID, 
            parsedAmount, 
            description, 
            paymentMethod, 
            initialStatus
        ];

        const result = await query(insertQuery, values);
        const newEscrow = result.rows[0];

        // --- Success Response ---
        return {
            statusCode: 201, 
            headers: HEADERS,
            body: JSON.stringify({
                message: `Escrow proposal created successfully with system Seller ID ${DEFAULT_SELLER_ID}, awaiting acceptance.`,
                escrow: newEscrow,
            }),
        };
    } catch (err) {
        console.error("Fatal error in buyerCreateEscrow:", err);
        
        let customError = err.message || "An unknown error occurred during escrow creation.";
        let status = 500;
        
        if (err.code === '23503') { 
            status = 404;
            customError = `The system Seller ID (${DEFAULT_SELLER_ID}) or Buyer ID is invalid.`;
        } else if (err.code === '22P02') { 
             status = 400;
             customError = "Invalid input format for ID or amount.";
        }

        return {
            statusCode: status,
            headers: HEADERS,
            body: JSON.stringify({
                error: customError,
            }),
        };
    }
};

File: netlifyfunctions/functions/createEscrow.py
import json
import os
import psycopg2
from http import cookies

def get_session_user(headers):
    """Helper to extract logged-in user ID and role from the session cookie."""
    cookie_header = headers.get("cookie") or headers.get("Cookie") or ""
    cookie_obj = cookies.SimpleCookie()
    cookie_obj.load(cookie_header)

    session_token = None
    if "session" in cookie_obj:
        session_token = cookie_obj["session"].value

    if not session_token:
        return None

    try:
        conn = psycopg2.connect(
            host=os.getenv("DB_HOST"),
            database=os.getenv("DB_NAME"),
            user=os.getenv("DB_USER"),
            password=os.getenv("DB_PASS"),
            sslmode="require"
        )
        cur = conn.cursor()
        cur.execute("""
            SELECT u.id, u.role 
            FROM sessions s 
            JOIN users u ON s.user_id = u.id 
            WHERE s.session_token = %s;
        """, (session_token,))
        user = cur.fetchone()
        cur.close()
        conn.close()
        if not user:
            return None
        return {"id": user[0], "role": user[1]}
    except Exception as e:
        print("Session check failed:", e)
        return None


def handler(event, context):
    """POST /.netlify/functions/createEscrow"""

    # Parse request body
    try:
        data = json.loads(event.get("body", "{}"))
    except Exception:
        return {
            "statusCode": 400,
            "body": json.dumps({"error": "Invalid JSON body"})
        }

    # Required fields
    seller_id = data.get("seller_id")
    amount = data.get("amount")
    payment_method = data.get("paymentMethod", "").strip()
    preferred_wallet = data.get("preferred_wallet", "").strip()
    agreement = data.get("agreement", "").strip()

    # Validate input
    if not amount or not payment_method:
        return {
            "statusCode": 400,
            "body": json.dumps({"error": "Missing required fields (amount, paymentMethod)"})
        }

    # Get session user
    user = get_session_user(event.get("headers", {}))
    if not user:
        return {"statusCode": 401, "body": json.dumps({"error": "Not authenticated"})}
    if user["role"] != "buyer":
        return {"statusCode": 403, "body": json.dumps({"error": "Only buyers can create escrows"})}

    # Connect to Neon DB
    try:
        conn = psycopg2.connect(
            host=os.getenv("DB_HOST"),
            database=os.getenv("DB_NAME"),
            user=os.getenv("DB_USER"),
            password=os.getenv("DB_PASS"),
            sslmode="require"
        )
        cur = conn.cursor()
    except Exception as e:
        return {
            "statusCode": 500,
            "body": json.dumps({"error": "Database connection failed", "details": str(e)})
        }

    try:
        cur.execute(
            """
            INSERT INTO escrows (buyer_id, seller_id, amount, payment_method, payment_details, agreement, status)
            VALUES (%s, %s, %s, %s, %s, %s, %s)
            RETURNING id;
            """,
            (
                user["id"],
                seller_id,
                amount,
                payment_method,
                preferred_wallet,
                agreement,
                "pending"
            )
        )
        escrow_id = cur.fetchone()[0]
        conn.commit()
        cur.close()
        conn.close()

        return {
            "statusCode": 200,
            "headers": {"Content-Type": "application/json"},
            "body": json.dumps({
                "message": "Escrow created successfully",
                "escrow_id": escrow_id
            })
        }

    except Exception as e:
        conn.rollback()
        cur.close()
        conn.close()
        return {
            "statusCode": 500,
            "body": json.dumps({"error": "Failed to create escrow", "details": str(e)})
        }
